src/Conversion/ONNXToTorch/NN/FlattenOp.cpp

```diff
@@ -0,0 +1,135 @@
+ *              tensor of bfloat16 type or tensor of 16-bit float values
+ *              tensor of 32-bit float or tensor of 64-bit float values
+ *              tensor of string type values or tensor of 1-bit signless
+ *              integer values or tensor of complex type with 32-bit float
+ *              elements values or tensor of complex type with 64-bit float
+ *              elements values or memref of any type values
+ *
+ */
+
+static Value createAtenFlattenOp(ConversionPatternRewriter &rewriter, 
+	Location loc, Value result, ValueTensorType resultType,
+	int64_t start_dim, int64_t end_dim, ONNXFlattenOp op1) {
+  auto ty = IntegerType::get(op1.getContext(), 64);
+  auto startDimInt = IntegerAttr::get(ty, (start_dim));
+  Value startDimConstInt = 
+	  rewriter.create<ConstantIntOp>(loc, startDimInt);
+  auto endDimInt = IntegerAttr::get(ty, (end_dim));
+  Value endDimConstInt = rewriter.create<ConstantIntOp>(loc, endDimInt);
+  return rewriter.create<AtenFlattenUsingIntsOp>(loc,
+                    resultType, result, startDimConstInt, endDimConstInt);
+}
+
+class ONNXFlattenOpToTorchLowering : public ConversionPattern {
+public:
+  ONNXFlattenOpToTorchLowering(TypeConverter &typeConverter,
+		  MLIRContext *ctx)
+      : ConversionPattern( typeConverter,
+		      ::mlir::ONNXFlattenOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+
+    Location loc = op->getLoc();
+    mlir::MLIRContext *context = op->getContext();
+    ONNXFlattenOp op1 = llvm::dyn_cast<ONNXFlattenOp>(op);
+
+    Value input = op1.input();
+    auto axisValue = op1.axis();       // ::mlir::IntegerAttr
+
+    auto inputShape = input.getType().cast<ShapedType>().getShape();
+    int64_t inputRank = inputShape.size();
+
+    TensorType resultTensorType =
+	    op->getResult(0).getType().cast<TensorType>();
+    auto resultTy = Torch::ValueTensorType::get(op1.getContext(),
+          resultTensorType.getShape(), resultTensorType.getElementType());
+
+    TensorType inputTensorType  = input.getType().cast<TensorType>();
+    auto inputTy =
+	 Torch::ValueTensorType::get(context, inputTensorType.getShape(),
+                    inputTensorType.getElementType());
+    auto inputTensor =
+	  rewriter.create<torch::TorchConversion::FromBuiltinTensorOp>(loc,
+                    inputTy, input);
+    Value result = inputTensor;
+
+    if (axisValue < 0)
+      return op->emitError("negative axis not supported");
+    
+    if (axisValue > 1) {
+      // flatten the region upto axis-1.
+      result = createAtenFlattenOp (rewriter, loc, result, resultTy, 0, 
+		      axisValue - 1, op1);
+      llvm::outs() << "Aten Flatten1 Op:   "
+	      << "\n" << result << "\n" << "\n";
+    }
+
+    // flatten the region from axis upwards.
+    result = createAtenFlattenOp (rewriter, loc, result, resultTy,
+		    axisValue, inputRank, op1);
```

Comment:
```suggestion
		    axisValue, inputRank, op1);
```
```suggestion
		    axisValue, inputRank - 1, op1);
```

Reply:
This magically works in an e2e test, but we would prefer compatibility with TorchMLIR. Here they utilize index addressing similar to Python.

Reply:
Resolved